activesupport/lib/active_support/tagged_logging.rb

```diff
@@ -114,29 +117,39 @@ def self.extended(base)
       end
     end
 
-    def self.new(logger)
-      logger = logger.clone
-
-      if logger.formatter
-        logger.formatter = logger.formatter.clone
+    def self.new(logger) # :nodoc:
+      if logger.is_a?(TaggedLogging)
+        logger.clone
       else
-        # Ensure we set a default formatter so we aren't extending nil!
-        logger.formatter = ActiveSupport::Logger::SimpleFormatter.new
+        logger.extend(TaggedLogging)
       end
+    end
+
+    def self.extended(base)
+      base.tag_processor = TagProcessor.new
+      base.extend(ActiveSupport::LogProcessor)
 
-      logger.formatter.extend Formatter
-      logger.extend(self)
+      base.processors << base.tag_processor
     end
 
-    delegate :push_tags, :pop_tags, :clear_tags!, to: :formatter
+    def initialize_clone(_)
+      self.tag_processor = TagProcessor.new
+      self.processors = [tag_processor]
+
+      super
+    end
+
+    delegate :push_tags, :pop_tags, :clear_tags!, to: :tag_processor
+    attr_accessor :tag_processor
 
     def tagged(*tags)
       if block_given?
-        formatter.tagged(*tags) { yield self }
+        tag_processor.tagged(*tags) { yield(self) }
       else
-        logger = ActiveSupport::TaggedLogging.new(self)
-        logger.formatter.extend LocalTagStorage
-        logger.push_tags(*formatter.current_tags, *tags)
+        logger = clone
+        logger.tag_processor.extend(LocalTagStorage)
```

Comment:
hi, could you also resolve the [issue 48787](https://github.com/rails/rails/issues/48787) ?
the `LocalTagStorage` is cause that `TagStack` is leaked across threads because it prevents the `Formatter` module (you change to class TagProcessor) to get the `IsolatedExecutionState TagStack` in each separate thread.
I wonder what the purpose of this `LocalTagStorage` anyway ? 

The issue 48787 also raise the confusion about when shared tagged-logs across threads and when not, could you also introduce a proposal to clear this confusion also in this PR ?
Perhaps we could introduce a `based_tagged` to store all share-tagged which will use acrcoss threads, so `TagStack` in each thread will be `based_tagged` + `IsolatedExecutionState TagStack`.

Reply:
Hey! I'd have to dig on #48787 and figure out the original intent to share tags across threads (thought from a quick glance this seems to be on purpose).
This PR is big enough and the mentioned issue isn't really related to the changes I'm proposing here, so this will potentially be changed in another patch.